minimu9-ahrs.1
.TH MINIMU9-AHRS 1
.SH NAME
minimu9-ahrs - read data from the Pololu MinIMU-9 over I²C
.SH SYNOPSIS
.B minimu9-ahrs [\fIOPTIONS\fP]
.SH DESCRIPTION
minimu9-ahrs is a program for reading data from the Pololu MinIMU-9 over I²C.
It supports MinIMU-9 versions v0, v1, v2, v3, v5, and v6.
The program can output the raw sensor data from the magnetometer, accelerometer,
and gyro or it can use that raw data to compute the orientation of the IMU.
.P
The program auto-detects what version of MinIMU-9 is connected.
.SS OPTIONS
.TP
\fB-h, --help\fP
produce help message
.TP
\fB-v, --version\fP
print version number
.TP
\fB-b, --i2c-bus\fP
I²C-bus the IMU is connected to (default: /dev/i2c-1)
.TP
\fB--mode \fIMODE\fR
Specify what algorithm to use.
.nf
\fBnormal\fP (default): Fuse compass and gyro.
\fBgyro-only\fP:  Use only gyro (drifts).
\fBcompass-only\fP:  Use only compass (noisy).
\fBraw\fP: Just print raw values from sensors.
.fi
.TP
\fB--output \fIFORMAT\fR
Specify how to output the orientation.
Has no effect if \fB--mode raw\fP is specified.
.nf
\fBmatrix\fP (default): Direction Cosine Matrix.
\fBquaternion\fP: Quaternion.
\fBeuler\fP: Euler angles (yaw, pitch, roll).

.SS COORDINATE SYSTEMS
There are two coordinate systems in use.
.P
The \fBground coordinate system\fP uses the following axes, in this order:
\fBNorth\fP, \fBEast\fP, and \fBDown\fP.
.P
The \fBbody coordinate system\fP represents the orientation of the IMU, and
uses the following axes, as labeled on the IMU:
\fBX\fP, \fBY\fP, and \fBZ\fP.

.SS RAW OUTPUT FORMAT

If \fB--mode raw\fP is specified, the format of the output will be
nine integers:
.IP
\fBMX MY MZ  AX AY AZ  GX GY GZ\fP
.P
where \fBMX MY MZ\fP is the raw reading from the magnetometer,
\fBAX AY AZ\fP is the raw reading from the accelerometer, and
\fBGX GY GZ\fP is the raw reading from the gyro.

.SS ORIENTATION OUTPUT FORMATS

Unless \fB--mode raw\fP is specified, the output format will be of the form:
.IP
\fIORIENTATION\fB AX AY AZ BX BY BZ\fR
.P
where the format of \fIORIENTATION\fP is determined by the
\fB--output\fP argument,
\fBAX AY AZ\fP is the scaled acceleration vector in units of 1 g, and
\fBBX BY BZ\fP is the scaled magnetic field vector, whose magnitude should
normally be close to 1.

.SS ORIENTATION AS A MATRIX

The default output format is \fB--output matrix\fP.
In this mode, a 3x3 transformation matrix
(also known as the direction cosine matrix) is sent to the standard output.
The matrix is sent in row-major order; the first three numbers on the line
are row 0, etc.
.P
The matrix is defined as the matrix that transforms column vectors from
the \fBbody coordinate system\fP to the \fBground coordinate system\fP.
Accordingly, its nine elements can be specified as dot products of the
unit vectors representing the two coordinate systems:
.IP
.nf
north·x  north·y  north·z
east·x   east·y   east·z
down·x   down·y   down·z
.fi

.SS ORIENTATION AS A QUATERNION
If \fB--output quaternion\fP is specified, the orientation is formatted
as a unit quaternion (four numbers).  It is a left quaternion that
transforms vectors from the \fBbody coordinate system\fP to the
\fBground coordinate system\fP.

.SS ORIENTATION AS EULER ANGLES
If \fB--output euler\fP is specified, the orintation is formatted as three
Euler angles in the following order: yaw, pitch, roll.
These are the standard angles used to represent the orientation of an airplane.
The angles are output in degrees.
.P
The Euler angles represent the amount of rotation needed in three
distinct steps to turn the ground coordinate system into the body
coordinate system.
.P
Imagine you start with a coordinate system that is equal to ground
coordinate system: north, east, down.
.P
Now rotate the coordinate system about its third axis
(still pointing down) by a certain angle that we will call the \fByaw\fP.
When viewed from above, the yaw rotation goes in the clockwise direction.
.P
Now rotate the coordinate system about its second axis by a certain
angle that we will call the \fBpitch\fP.  A positive pitch makes
the first axis of the coordinate system go up.  A pitch of 90.0
would make the first axis point straight up, while a pitch of -90.0
would make the first axis point straight down.
.P
Now rotate the coordinate system about is first axis by a certain
angle that we will call the \fBroll\fP.
.P
The coordinate system you now have is equal to the
\fBbody coordinate system\fP (x, y, z).

.SH FILES
.TP
~/.minimu9-ahrs-cal
Calibration file for the magnetomer, needed unless \fB--mode raw\fP is specified.
This file should be a one-line file with 6 integers separated by spaces:
minimum x, maximum x, minimum y, maximum y, minimum z, maximum z.
These numbers specify the linear mapping from the raw ellipsoid to
the unit sphere. For example, if "minimum x" is -414, it means that a
magnetometer reading of -414 on the X axis will get mapped to -1.0 when
the readings are scaled.  Run \fBminimu9-ahrs-calibrate\fP to generate this
file.
.TP
~/.minimu9-ahrs
Configuration file that lets you set default values for some of the command-line
options.  Every line should have the format \fBKEY=VALUE\fP.  Currently, the
only supported option is i2c-bus.  To set the default I²C bus to /dev/i2c-1, you
would write:
.RS 2
.IP
i2c-bus=/dev/i2c-1
.SH AUTHOR
.nf
David Grayson <davidegrayson@gmail.com>
http://www.github.com/DavidEGrayson/
.fi
.SH SEE ALSO
minimu9-ahrs-calibrate(1), minimu9-ahrs-calibrator(1)


minimu9-ahrs.cpp
// For efficiency in compiling, this is the only file that uses the
// Eigen lirbary and the 'vector' type we made from it.

#include "vector.h"
#include "version.h"
#include "prog_options.h"
#include "minimu9.h"
#include "exceptions.h"
#include "pacer.h"
#include <iostream>
#include <iomanip>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include <pthread.h>
#include <system_error>
#include <chrono>

// TODO: print warning if accelerometer magnitude is not close to 1 when starting up

// An Euler angle could take 8 chars: -234.678, but usually we only need 6.
float field_width = 6;
#define FLOAT_FORMAT std::fixed << std::setprecision(3) << std::setw(field_width)

std::ostream & operator << (std::ostream & os, const vector & vector)
{
  return os << FLOAT_FORMAT << vector(0) << ' '
            << FLOAT_FORMAT << vector(1) << ' '
            << FLOAT_FORMAT << vector(2);
}

std::ostream & operator << (std::ostream & os, const matrix & matrix)
{
  return os << (vector)matrix.row(0) << ' '
            << (vector)matrix.row(1) << ' '
            << (vector)matrix.row(2);
}

std::ostream & operator << (std::ostream & os, const quaternion & quat)
{
  return os << FLOAT_FORMAT << quat.w() << ' '
            << FLOAT_FORMAT << quat.x() << ' '
            << FLOAT_FORMAT << quat.y() << ' '
            << FLOAT_FORMAT << quat.z();
}

typedef void rotation_output_function(quaternion & rotation);

void output_quaternion(quaternion & rotation)
{
  std::cout << rotation;
}

void output_matrix(quaternion & rotation)
{
  std::cout << rotation.toRotationMatrix();
}

void output_euler(quaternion & rotation)
{
  std::cout << (vector)(rotation.toRotationMatrix().eulerAngles(2, 1, 0)
                        * (180 / M_PI));
}

void stream_raw_values(imu & imu)
{
  imu.enable();
  while(1)
  {
    imu.read_raw();
    printf("%7d %7d %7d  %7d %7d %7d  %7d %7d %7d\n",
           imu.m[0], imu.m[1], imu.m[2],
           imu.a[0], imu.a[1], imu.a[2],
           imu.g[0], imu.g[1], imu.g[2]
      );
    usleep(20*1000);
  }
}

//! Uses the acceleration and magnetic field readings from the compass
// to get a noisy estimate of the current rotation matrix.
// This function is where we define the coordinate system we are using
// for the ground coords:  North, East, Down.
matrix rotation_from_compass(const vector & acceleration, const vector & magnetic_field)
{
  vector down = -acceleration;     // usually true
  vector east = down.cross(magnetic_field); // actually it's magnetic east
  vector north = east.cross(down);

  east.normalize();
  north.normalize();
  down.normalize();

  matrix r;
  r.row(0) = north;
  r.row(1) = east;
  r.row(2) = down;
  return r;
}

typedef void fuse_function(quaternion & rotation, float dt, const vector & angular_velocity,
  const vector & acceleration, const vector & magnetic_field);

void fuse_compass_only(quaternion & rotation, float dt, const vector& angular_velocity,
  const vector & acceleration, const vector & magnetic_field)
{
  (void)dt;
  (void)angular_velocity;
  (void)acceleration;
  (void)magnetic_field;
  // Implicit conversion of rotation matrix to quaternion.
  rotation = rotation_from_compass(acceleration, magnetic_field);
}

// Uses the given angular velocity and time interval to calculate
// a rotation and applies that rotation to the given quaternion.
// w is angular velocity in radians per second.
// dt is the time.
void rotate(quaternion & rotation, const vector & w, float dt)
{
  // Multiply by first order approximation of the
  // quaternion representing this rotation.
  rotation *= quaternion(1, w(0)*dt/2, w(1)*dt/2, w(2)*dt/2);
  rotation.normalize();
}

void fuse_gyro_only(quaternion & rotation, float dt, const vector & angular_velocity,
  const vector & acceleration, const vector & magnetic_field)
{
  (void)acceleration;
  (void)magnetic_field;
  rotate(rotation, angular_velocity, dt);
}

void fuse_default(quaternion & rotation, float dt, const vector & angular_velocity,
  const vector & acceleration, const vector & magnetic_field)
{
  vector correction = vector(0, 0, 0);

  if (fabs(acceleration.norm() - 1) <= 0.3)
  {
    // The magnetidude of acceleration is close to 1 g, so
    // it might be pointing up and we can do drift correction.

    const float correction_strength = 1;

    matrix rotation_compass = rotation_from_compass(acceleration, magnetic_field);
    matrix rotation_matrix = rotation.toRotationMatrix();

    correction = (
      rotation_compass.row(0).cross(rotation_matrix.row(0)) +
      rotation_compass.row(1).cross(rotation_matrix.row(1)) +
      rotation_compass.row(2).cross(rotation_matrix.row(2))
      ) * correction_strength;

  }

  rotate(rotation, angular_velocity + correction, dt);
}

void ahrs(imu & imu, fuse_function * fuse, rotation_output_function * output)
{
  imu.load_calibration();
  imu.enable();
  imu.measure_offsets();

  // The quaternion that can convert a vector in body coordinates
  // to ground coordinates when it its changed to a matrix.
  quaternion rotation = quaternion::Identity();

  // Set up a timer that expires every 20 ms.
  pacer loop_pacer;
  loop_pacer.set_period_ns(20000000);

  auto start = std::chrono::steady_clock::now();
  while(1)
  {
    auto last_start = start;
    start = std::chrono::steady_clock::now();
    std::chrono::nanoseconds duration = start - last_start;
    float dt = duration.count() / 1e9;
    if (dt < 0){ throw std::runtime_error("Time went backwards."); }

    vector angular_velocity = imu.read_gyro();
    vector acceleration = imu.read_acc();
    vector magnetic_field = imu.read_mag();

    fuse(rotation, dt, angular_velocity, acceleration, magnetic_field);

    output(rotation);
    std::cout << "  " << acceleration << "  " << magnetic_field << std::endl;

    loop_pacer.pace();
  }
}

int main_with_exceptions(int argc, char **argv)
{
  prog_options options = get_prog_options(argc, argv);

  if(options.show_help)
  {
    print_command_line_options_desc();
    std::cout << "For more information, run: man minimu9-ahrs" << std::endl;
    return 0;
  }

  if (options.show_version)
  {
    std::cout << VERSION << std::endl;
    return 0;
  }

  // Decide what sensors we want to use.
  sensor_set set;
  set.mag = set.acc = set.gyro = true;

  minimu9::comm_config config = minimu9::auto_detect(options.i2c_bus_name);

  sensor_set missing = set - minimu9::config_sensor_set(config);
  if (missing)
  {
    if (missing.mag)
    {
      std::cerr << "Error: No magnetometer found." << std::endl;
    }
    if (missing.acc)
    {
      std::cerr << "Error: No accelerometer found." << std::endl;
    }
    if (missing.gyro)
    {
      std::cerr << "Error: No gyro found." << std::endl;
    }
    std::cerr << "Error: Needed sensors are missing." << std::endl;
    return 1;
  }

  config = minimu9::disable_redundant_sensors(config, set);

  minimu9::handle imu;
  imu.open(config);

  rotation_output_function * output;

  // Figure out the output mode.
  if (options.output_mode == "matrix")
  {
    output = &output_matrix;
  }
  else if (options.output_mode == "quaternion")
  {
    output = &output_quaternion;
  }
  else if (options.output_mode == "euler")
  {
    field_width += 2;  // See comment above for field_width.
    output = &output_euler;
  }
  else
  {
    std::cerr << "Unknown output mode '" << options.output_mode << "'" << std::endl;
    return 1;
  }

  // Figure out the basic operating mode and start running.
  if (options.mode == "raw")
  {
    stream_raw_values(imu);
  }
  else if (options.mode == "gyro-only")
  {
    ahrs(imu, &fuse_gyro_only, output);
  }
  else if (options.mode == "compass-only")
  {
    ahrs(imu, &fuse_compass_only, output);
  }
  else if (options.mode == "normal")
  {
    ahrs(imu, &fuse_default, output);
  }
  else
  {
    std::cerr << "Unknown mode '" << options.mode << "'" << std::endl;
    return 1;
  }
  return 0;
}

int main(int argc, char ** argv)
{
  try
  {
    main_with_exceptions(argc, argv);
  }
  catch(const std::system_error & error)
  {
    std::string what = error.what();
    const std::error_code & code = error.code();
    std::cerr << "Error: " << what << " (" << code << ")" << std::endl;
    return 2;
  }
  catch(const std::exception & error)
  {
    std::cerr << "Error: " << error.what() << std::endl;
    return 9;
  }
}
